src/frontend/Typechecker.ml

```diff
@@ -0,0 +1,1520 @@
+  else if Utils.is_unnormalized_distribution id.name then
+    Semantic_error.udf_is_unnormalized_fn id.id_loc id.name |> error
+  else if Env.mem tenv id.name then
+    (* adapt for overloading later *)
+    Semantic_error.ident_in_use id.id_loc id.name |> error
+  else ()
+
+(** Checks that a variable/function name:
+  - a function/identifier does not have the _lupdf/_lupmf suffix
+  - is not already in use (for now)
+*)
+let verify_name_fresh tenv id ~is_udf =
+  let f =
+    if is_udf then verify_name_fresh_udf tenv else verify_name_fresh_var tenv
+  in
+  List.iter ~f (distribution_name_variants id)
+
+let is_of_compatible_return_type rt1 srt2 =
+  UnsizedType.(
+    match (rt1, srt2) with
+    | Void, NoReturnType
+     |Void, Incomplete Void
+     |Void, Complete Void
+     |Void, AnyReturnType ->
+        true
+    | ReturnType UReal, Complete (ReturnType UInt) -> true
+    | ReturnType UComplex, Complete (ReturnType UReal) -> true
+    | ReturnType UComplex, Complete (ReturnType UInt) -> true
+    | ReturnType rt1, Complete (ReturnType rt2) -> rt1 = rt2
+    | ReturnType _, AnyReturnType -> true
+    | _ -> false)
+
+(* -- Expressions ------------------------------------------------- *)
+let check_ternary_if loc pe te fe =
+  match
+    (pe.emeta.type_, UnsizedType.common_type (te.emeta.type_, fe.emeta.type_))
+  with
+  | UInt, Some type_ when not (UnsizedType.is_fun_type type_) ->
+      mk_typed_expression
+        ~expr:(TernaryIf (pe, te, fe))
+        ~ad_level:(expr_ad_lub [pe; te; fe])
+        ~type_ ~loc
+  | _, _ ->
+      Semantic_error.illtyped_ternary_if loc pe.emeta.type_ te.emeta.type_
+        fe.emeta.type_
+      |> error
+
+let check_binop loc op le re =
+  let rt =
+    [le; re] |> get_arg_types
+    |> Stan_math_signatures.operator_stan_math_return_type op
+  in
+  match rt with
+  | Some (ReturnType type_) ->
+      mk_typed_expression
+        ~expr:(BinOp (le, op, re))
+        ~ad_level:(expr_ad_lub [le; re])
+        ~type_ ~loc
+  | _ ->
+      Semantic_error.illtyped_binary_op loc op le.emeta.type_ re.emeta.type_
+      |> error
+
+let check_prefixop loc op te =
+  let rt =
+    Stan_math_signatures.operator_stan_math_return_type op [arg_type te]
+  in
+  match rt with
+  | Some (ReturnType type_) ->
+      mk_typed_expression
+        ~expr:(PrefixOp (op, te))
+        ~ad_level:(expr_ad_lub [te])
+        ~type_ ~loc
+  | _ -> Semantic_error.illtyped_prefix_op loc op te.emeta.type_ |> error
+
+let check_postfixop loc op te =
+  let rt =
+    Stan_math_signatures.operator_stan_math_return_type op [arg_type te]
+  in
+  match rt with
+  | Some (ReturnType type_) ->
+      mk_typed_expression
+        ~expr:(PostfixOp (te, op))
+        ~ad_level:(expr_ad_lub [te])
+        ~type_ ~loc
+  | _ -> Semantic_error.illtyped_postfix_op loc op te.emeta.type_ |> error
+
+let check_variable cf loc tenv id =
+  match Env.find tenv (Utils.stdlib_distribution_name id.name) with
```

Comment:
So `tenv` only contains normalized distribution names. Would it be okay to add the unnormalized names too? Then you wouldn't need to reconstruct the actual type with `Fun_kind.suffix_from_name`.